### PR TITLE
compilers: cpp: improve libc++ vs libstdc++ detection (again)

### PR DESCRIPTION
**Pull request by thesamesam**
The previous approach wasn't great because you couldn't control what the detected C++ stdlib impl was. We just had a preference list we tweaked the searched order for per OS. That doesn't work great for e.g. Gentoo with libc++ or Gentoo Prefix on macOS where we might be using libstdc++ even though the host is libc++.

Jonathan Wakely, the libstdc++ maintainer, gave a helpful answer on how to best detect libc++ vs libstdc++ via macros on SO [0]. Implement it.

TL;DR: Use <version> from C++20 if we can, use <ciso646> otherwise. Check for _LIBCPP_VERSION as libstdc++ doesn't always define a macro.

[0] https://stackoverflow.com/a/31658120

### CODE REVIEW

File: mesonbuild/compilers/cpp.py [hunk 1]
```diff
@@ -203,17 +203,19 @@ def language_stdlib_only_link_flags(self, env: Environment) -> T.List[str]:
...
-        for lib in search_order:
-            if self.find_library(lib, env, []) is not None:
-                return search_dirs + [f'-l{lib}']
+        # https://stackoverflow.com/a/31658120
+        try:
+            # If we can, use C++20's <version>.
+            self._find_best_cpp_std('c++20')
+            header = 'version'
+        except MesonException:
+            header = 'ciso646'
```
Comment (dcbaker) [inline mesonbuild/compilers/cpp.py:212]: Does it even matter if C++ has c++20 here, or can we just do something like (not sure what the method actually is):
```python
header ='version' if self.has_header('<version>') else 'cisco646'
```

Comment (thesamesam) [reply mesonbuild/compilers/cpp.py:212]: Nah, we can fall back fine if we do that in theory. Let me try it.. Thanks!

**Review by dcbaker — APPROVED** [review]

File: mesonbuild/compilers/cpp.py (hunk 1 above)
Comment (eli-schwartz) [reply mesonbuild/compilers/cpp.py]: Nice idea. That also means we can do a preprocess check for the header presence instead of a compile check to see if the -std option is a valid flag. In theory this is a bit faster.
